tests/unit/exception_logging_test.cc

```diff
@@ -0,0 +1,142 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Copyright 2016 ScyllaDB
+ */
+
+#define BOOST_TEST_MODULE core
+
+#include <boost/test/included/unit_test.hpp>
+#include <seastar/util/log.hh>
+#include <ostream>
+
+
+using namespace seastar;
+
+// a class which is not derived from std::exception
+// to play the part of the unknown object in the logging
+// function.
+class unknown_obj {
+    sstring _message;
+public:
+    unknown_obj(std::string message) : _message(message) {}
+};
+
+// This functions generates an exception chain nesting_level+1 deep
+// for each nesting level it throws one of two types of objects, an
+// unknown (non std::exception) object or a runtime error which is
+// derived from std::exception, it chooses the type of thrown object
+// according to the bit in the `nesting_level` place in the
+// `tests_instance` paramter or in other words according to:
+// bool(test_instance & (1<<nesting_level))
+void exception_generator(uint32_t test_instance, int nesting_level) {
+    try {
+        if (nesting_level > 0) {
+            exception_generator(test_instance>>1, nesting_level-1);
+        }
+    } catch(...) {
+        auto msg = format("Exception Level {}", nesting_level);
+        if(test_instance&1) {
+            // Throw a non std::exception derived type
+            std::throw_with_nested(unknown_obj(msg));
+        } else {
+            std::throw_with_nested(std::runtime_error(msg));
+        }
+    }
+    if (nesting_level == 0) {
+        if (test_instance & 1) {
+            throw unknown_obj(format("Exception Level {}", nesting_level));
+        } else {
+            throw std::runtime_error(format("Exception Level {}", nesting_level));
+        }
+    }
+}
+
+// This function generates the expected logging output string of an exception
+// thrown by the  exception generator function with a specific output.
+std::string exception_generator_str(uint32_t test_instance,int nesting_level) {
```

Comment:
I understand that point, for completeness, here are the strings that are produces by the run of this generator test:
```
unknown_obj (no extra info)
std::runtime_error (Exception Level 0)
std::_Nested_exception<unknown_obj>: unknown nested type: unknown_obj (no extra info)
std::_Nested_exception<std::runtime_error> (Exception Level 1): unknown_obj (no extra info)
std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
std::_Nested_exception<std::runtime_error> (Exception Level 1): std::runtime_error (Exception Level 0)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<unknown_obj>: unknown nested type: unknown_obj (no extra info)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<unknown_obj>: unknown nested type: unknown_obj (no extra info)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<std::runtime_error> (Exception Level 1): unknown_obj (no extra info)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<std::runtime_error> (Exception Level 1): unknown_obj (no extra info)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<std::runtime_error> (Exception Level 1): std::runtime_error (Exception Level 0)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<std::runtime_error> (Exception Level 1): std::runtime_error (Exception Level 0)
```

Reply:
Thanks. I already committed this series, so if we want to improve it will need to be a followup patch.
In these examples you can see what I asked about the string "unknown exception type"... Check out: 
```
std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
```
What does "unknown nested type" refer to? The message said it's std::nested_exception<unknown_obj> so we know it's an unknown_obj, and, as the next part of the message says - also an std::runtime_error. So what has "unknown type"?

Reply:
Now I see the point, maybe I should change it to `non exception derived type (no extra info)`.
I can also just skip this entirely.
I am inclining towards that it is not that important but I could add another patch if it will bother someone.

Reply:
If you really want that "no extra info" on unknown exception types, you should - I think - at least be consistent and add it here like in non-nested exceptions. I.e., the above message will become:
```
std::_Nested_exception<unknown_obj> (no extra info): std::runtime_error (Exception Level 0)
```
i.e., no "unknown type" message (which feels wrong to me, the type is known), no "non exception derived type", just a "no extra info" exactly like in the other catch(...).

I still don't know what this "no extra info" message adds or clarifies, though. But at least it's not misleading like (I think) "unknown type" is misleading.